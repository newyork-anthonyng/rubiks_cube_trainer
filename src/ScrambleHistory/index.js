import React, { Fragment } from "react";
import PropTypes from "prop-types";

const ScrambleHistory = ({ history }) => (
  <Fragment>
    <h2>Scramble History</h2>
    {history.length === 0 ? (
      <p>No scramble history available</p>
    ) : (
      <ul>
        {history.map(({ scramble, difficulty }) => (
          <li key={scramble}>
            <span>{difficulty}: </span>
            <span>{scramble}</span>
          </li>
        ))}
      </ul>
    )}
  </Fragment>
);

ScrambleHistory.defaultProps = {
  history: []
};

ScrambleHistory.propTypes = {
  history: PropTypes.arrayOf(
    PropTypes.shape({
      scramble: PropTypes.string,
      difficulty: PropTypes.number
    })
  )
};

export default ScrambleHistory;
