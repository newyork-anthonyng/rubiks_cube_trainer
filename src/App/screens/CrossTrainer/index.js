import React, { Component } from "react";
import PropTypes from "prop-types";

import ScrambleHistory from "./components/ScrambleHistory";

import getRandomSolve from "rubiks-cross-trainer";
import createEmptyArray from "create-empty-array";

class CrossTrainer extends Component {
  constructor(props) {
    super(props);

    this.optionsArray = createEmptyArray(8);
    this.initialDifficulty = 1;

    this.state = {
      difficulty: this.initialDifficulty,
      scramble: props.initialScramble || getRandomSolve(this.initialDifficulty),
      scrambleHistory: []
    };
  }

  getRandomSolve = () => {
    this.setState(prevState => ({
      scramble: getRandomSolve(prevState.difficulty),
      scrambleHistory: [
        {
          difficulty: prevState.difficulty,
          scramble: prevState.scramble
        }
      ].concat(prevState.scrambleHistory)
    }));
  };

  handleSelectChange = e => {
    this.setState({
      difficulty: e.target.value
    });
  };

  render() {
    const containerStyle = {
      display: "flex",
      flexDirection: "column"
    };

    const headerStyle = {
      textAlign: "center"
    };

    const formStyle = {
      alignItems: "center",
      display: "flex",
      flexDirection: "column",
      justifyContent: "center"
    };

    const selectStyle = {
      marginBottom: "1rem"
    };

    return (
      <div style={containerStyle}>
        <h1 style={headerStyle}>{this.state.scramble}</h1>

        <div style={formStyle}>
          <div style={selectStyle}>
            <span>Number of moves to generate cross:</span>
            <select
              value={this.state.difficulty}
              onChange={this.handleSelectChange}
            >
              {this.optionsArray.map((_, index) => (
                <option key={index} value={index + 1}>
                  {index + 1}
                </option>
              ))}
            </select>
          </div>
          <button onClick={this.getRandomSolve}>Get random solve</button>
        </div>

        <ScrambleHistory history={this.state.scrambleHistory} />
      </div>
    );
  }
}

CrossTrainer.propTypes = {
  initialScramble: PropTypes.string
};

export default CrossTrainer;
