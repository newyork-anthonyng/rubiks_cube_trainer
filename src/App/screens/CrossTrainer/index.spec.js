jest.mock("rubiks-cross-trainer", () => {
  let count = 0;
  return jest.fn(() => `SOME_RANDOM_SCRAMBLE_${count}`);
});
import getRandomSolve from "rubiks-cross-trainer";

import React from "react";
import CrossTrainer from "./";
import ScrambleHistory from "./components/ScrambleHistory";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

beforeEach(() => {
  getRandomSolve.mockClear();
});

it("should render correctly", () => {
  const wrapper = shallow(<CrossTrainer />);

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should get new scramble when clicking on button", () => {
  const wrapper = shallow(<CrossTrainer />);
  getRandomSolve.mockClear();

  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  expect(getRandomSolve).toHaveBeenCalledTimes(1);
  expect(getRandomSolve.mock.calls[0][0]).toEqual(1);
});

it("should update scramble history when clicking on button", () => {
  const wrapper = shallow(<CrossTrainer />);
  getRandomSolve.mockClear();

  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  expect(wrapper.find(ScrambleHistory).props()).toMatchSnapshot();
});

it("should use new count value when changing select field", () => {
  const wrapper = shallow(<CrossTrainer />);
  getRandomSolve.mockClear();

  const selectEl = wrapper.find("select");
  const newValue = 5;
  selectEl.simulate("change", { target: { value: newValue } });

  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  expect(getRandomSolve).toHaveBeenCalledTimes(1);
  expect(getRandomSolve.mock.calls[0][0]).toEqual(newValue);
});

it("should use initial scramble if provided", () => {
  const wrapper = shallow(<CrossTrainer initialScramble="R2 U" />);

  expect(getRandomSolve).not.toHaveBeenCalled();
  expect(toJSON(wrapper)).toMatchSnapshot();
});
