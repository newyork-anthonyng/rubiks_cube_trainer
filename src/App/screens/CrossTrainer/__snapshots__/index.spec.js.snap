// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    SOME_RANDOM_SCRAMBLE_0
  </h1>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <div
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
    >
      <span>
        Number of moves to generate cross:
      </span>
      <select
        onChange={[Function]}
        value={1}
      >
        <option
          key="0"
          value={1}
        >
          1
        </option>
        <option
          key="1"
          value={2}
        >
          2
        </option>
        <option
          key="2"
          value={3}
        >
          3
        </option>
        <option
          key="3"
          value={4}
        >
          4
        </option>
        <option
          key="4"
          value={5}
        >
          5
        </option>
        <option
          key="5"
          value={6}
        >
          6
        </option>
        <option
          key="6"
          value={7}
        >
          7
        </option>
        <option
          key="7"
          value={8}
        >
          8
        </option>
      </select>
    </div>
    <button
      onClick={[Function]}
    >
      Get random solve
    </button>
  </div>
  <ScrambleHistory
    history={Array []}
  />
</div>
`;

exports[`should update scramble history when clicking on button 1`] = `
Object {
  "history": Array [
    Object {
      "difficulty": 1,
      "scramble": "SOME_RANDOM_SCRAMBLE_0",
    },
  ],
}
`;

exports[`should use initial scramble if provided 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    R2 U
  </h1>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <div
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
    >
      <span>
        Number of moves to generate cross:
      </span>
      <select
        onChange={[Function]}
        value={1}
      >
        <option
          key="0"
          value={1}
        >
          1
        </option>
        <option
          key="1"
          value={2}
        >
          2
        </option>
        <option
          key="2"
          value={3}
        >
          3
        </option>
        <option
          key="3"
          value={4}
        >
          4
        </option>
        <option
          key="4"
          value={5}
        >
          5
        </option>
        <option
          key="5"
          value={6}
        >
          6
        </option>
        <option
          key="6"
          value={7}
        >
          7
        </option>
        <option
          key="7"
          value={8}
        >
          8
        </option>
      </select>
    </div>
    <button
      onClick={[Function]}
    >
      Get random solve
    </button>
  </div>
  <ScrambleHistory
    history={Array []}
  />
</div>
`;
